### PR TITLE
Feature/1618 node specific parameters

### PR DESCRIPTION
**Pull request by lchambon**
**PR description**

This draft PR introduces a first attempt to address issue #1618 regarding node-specific parameters, implementing the design discussions from issues #1620, #1627 and #1628.

**Summary of changes**

Researcher can now tag parameters in a training plan as persistent and/or private:

- private: parameters that are never shared with the researcher (i.e., never aggregated).
- persistent: parameters that are saved locally on the node after each round.

In simple scenarios, this implementation works correctly for both Scikit-learn and PyTorch examples.
After fixing a small bug in breakpoint loading, breakpoints also appear to work correctly.

**Open questions & discussion points**

About Differential Privacy:

- Currently, OPACUS modifies model parameter names by adding a `"_module"` prefix, which breaks the current implementation because parameters cannot be loaded by the original names specified in `tag_parameters`.
- Should DP be applied to private parameters that are never shared to researcher? My intuition is no, but it’s unclear how this affects implementation. What are the consequences if the set of parameters subject to DP is smaller than expected
- While modifying the `after_training_params` function in `TorchTrainingPlan`, it seems that both SecAgg and DP cannot be applied simultaneously:

```
def after_training_params(self, flatten: bool = False) -> Dict[str, torch.Tensor]:
    ...

    # Run (optional) DP controller adjustments as well.
    params = self._dp_controller.after_training(params)
    if flatten:
        params = self._model.flatten(
            exclude_buffers=not self._share_persistent_buffers,
            private_params=self._private_params,
        )
    return params
 ```
 At least, DP seems to be ignored if SecAgg is applied in the `if flatten` statement.

About Declearn:

- How does the private parameter concept impact Declearn and more complex optimizers?
- Currently, using Scaffold triggers warnings, which require investigation.

About Secure Aggregation:

- The different use cases tested with secure aggregation (PyTorch and Scikit-learn) seem to work fine.
- However, the `secure_aggregation` function in `Aggregator` does not appear to be used anywhere. Should it still be updated to reflect the new parameter access logic, or can it be deprecated?

**Missing**

- [x] Implement unit tests.
- [x] Update documentation.
- [x] Create e2e test

### CODE REVIEW

**Comment by mvesin** [comment]
Great work @lchambon !

The open questions you raise are all very relevant and advanced points - may need some in depth team discussion :-)

My 2 cents at first read, HTH

> Should DP be applied to private parameters that are never shared to researcher? My intuition is no, but it’s unclear 
> how this affects implementation. What are the consequences if the set of parameters subject to DP is smaller than 
> expected

We need a DP expert on this :-)

Probably local DP and central DP are 2 different cases. My first intuition would be

1. in local DP: as it is applied in each batch training, the private parameters value would influence non-private parameters in next batch training - so it should be applied to avoid breaking model privacy.

AI assistants suggest that it is often not needed, as this case is often handled by DP applied to shared layers in next rounds. Leakage that may result from non-clipped gradients with DP are general issues, not introduced by private layers. Main risk would come from indirect leakage from the private layers (eg reuse local model for inference, distillation; client side attacks) which are probably out of our scope.

All in all, applying noise to all layers would be a reasonable first implementation choice (simple to implement and conservative) though sub optimal (noising private layers in many scenario where it is not needed).

More advanced (future) implementation may not noise private layers by default, with a flag to change behaviour.

2. in central DP: as it is applied after training, it's applying noise to params that won't be shared with the researcher. Yet, if these params are persistent it matters. In that case these params will influence non-private layers in next rounds. But then they will be subject to the local/central DP policy of the node. So it may not be necessary to apply central DP to the private layer. So clip + noise only shared layers is OK.

AI assistant suggest that noising private layers is probably not needed. Main risks are indirect leakage mentioned for local DP (probably out of our scope).

So we may choose not noise the private layer if simple enough to implement.
We may also noise them as first simple implementation (suboptimal utility) - and maybe document this in user documentation .
Or have optional param to do it in more advanced implementation.

>  While modifying the after_training_params function in TorchTrainingPlan, it seems that both SecAgg and DP cannot be applied simultaneously (at least, DP seems to be ignored if SecAgg is applied). Is this intended?

Sorry I could not find that. Can you point to the code where it happens to remind us ?

>  The different use cases tested with secure aggregation (PyTorch and Scikit-learn) seem to work fine.
>  However, the secure_aggregation function in Aggregator does not appear to be used anywhere. Should it still be 
> updated to reflect the new parameter access logic, or can it be deprecated?

Good point, I could not find any use of this function too.
@srcansiz had begun a refactor of secagg that could not be completed - maybe linked to this ?

HTH !

File: fedbiomed/node/round.py [hunk 1]
```diff
@@ -1093,3 +1168,61 @@ def _split_train_and_test_data(self, test_ratio: float = 0) -> DataManager:
...
+        Args:
+            initial_model_weights: initial values of all model parameters (before any training)
+            weights_from_researcher: parameters received from the researcher
+            persistent_model_weights: locally stored parameters saved from previous rounds
+
+        Returns:
+            the complete set of parameters to be loaded into the model before training
+        """
+        # Deal with the case where persistent_model_weights is None
+        persistent_model_weights = persistent_model_weights or {}
```
Comment (sharkovsky) [inline fedbiomed/node/round.py:1201]: Handle edge cases where the weight itself evaluates to false (e.g. empty dict, but maybe the weight was actually an empty dict. Or the weight we a single scalar boolean set to False)

```suggestion
        persistent_model_weights = persistent_model_weights if persistent_model_weights is not None else {}
```

File: fedbiomed/node/round.py [hunk 2]
```diff
@@ -1093,3 +1168,61 @@ def _split_train_and_test_data(self, test_ratio: float = 0) -> DataManager:
...
+
+        # Start from parameters received by researcher
+        full_params = dict(weights_from_researcher)
```
Comment (sharkovsky) [inline fedbiomed/node/round.py:1204]: why is this needed? Isn't `weights_from_researcher` already a dict?

File: fedbiomed/node/round.py [hunk 3]
```diff
@@ -1093,3 +1168,61 @@ def _split_train_and_test_data(self, test_ratio: float = 0) -> DataManager:
...
+        local_only = self.training_plan.filter_model_params_by_tags(
+            initial_model_weights,
+            required_tags={"local"},
+            forbidden_tags={"persistent"},
+        )
+
+        for name in local_only.keys():
+            full_params[name] = initial_model_weights[name]
+
+        return full_params
```
Comment (sharkovsky) [inline fedbiomed/node/round.py:1228]: please DOUBLE-CHECK what I'm saying, but I think that all this code could be achieved with a one-liner

```suggestion
        return {**initial_model_weights, **persistent_model_weights, ****weights_from_researcher}
```

It should achieve the following behaviour:
- local + persistent parameters are overwritten with their persistent value
- persistent-only parameters are overwritten with their aggregated value
- automatic first-round handling: since the local+persistent parameters will not be in either the persistent or researcher dictionaries, they will automatically be randomly initialized

Example
```python
In [14]: weights_from_researcher = {'agg': 3, 'agg4': 4, 'perst': 1}
In [15]: persistent_model_weights = {'perst': 11, 'perst_local': 2}
In [16]: initial_model_weights = {'local': 0, 'perst': 11, 'perst_local': 222, 'agg': 33, 'agg4': 44}

In [17]: {**initial_model_weights, **persistent_model_weights, **weights_from_researcher}
Out[17]: {'local': 0, 'perst': 1, 'perst_local': 2, 'agg': 3, 'agg4': 4}
```

File: docs/user-guide/researcher/training-plan.md [hunk 4]
```diff
@@ -393,4 +393,103 @@ model_args = {
...
+        pass
+
+    def init_optimizer(self, optimizer_args):
+        pass
+
+    def init_dependencies(self):
+        pass
+```
+
+### What happens on the node side during training
```
Comment (sharkovsky) [inline docs/user-guide/researcher/training-plan.md:473]:
```suggestion
### Handling of parameters tagged as `local` or `persistent` during training
```

File: docs/user-guide/researcher/training-plan.md [hunk 5]
```diff
@@ -393,4 +393,103 @@ model_args = {
...
+
+- Parameters tagged as both `local` and `persistent` are restored from the values saved at the end of the previous round. If no previous round exists (i.e., the first round), they are initialized from the model's initial weights as defined in `init_model`.
+- Parameters tagged as `local` only (without `persistent`) are always reinitialized from the model's initial weights at each round, regardless of any training that may have occurred in previous rounds.
+- All other parameters are taken as-is from the researcher-provided aggregated weights.
+
+**At the end of a round**, the node saves the current values of all parameters tagged as `persistent`. These saved values will be loaded and restored at the beginning of the next round.
+
+**When sending parameters back to the researcher**, parameters tagged as `local` are excluded from the payload. The researcher therefore never sees or aggregates them, and they play no role in the global model update.
+
```
Comment (sharkovsky) [inline docs/user-guide/researcher/training-plan.md:486]: Please add some explanation for the researcher:
- the `local` parameters are not available in the `experiment.aggregated_params()`, nor in the `training_replies`
- however, the `local` parameters are still there in `experiment.training_plan().get_model_params()` and inside `experiment.training_plan().model()`. But clarify that these values are taken from `init_model`, and do not correspond to what the node will see in the next round.

File: fedbiomed/common/models/_sklearn.py [hunk 6]
```diff
@@ -121,10 +125,21 @@ def get_weights(
                 f"{ErrorNumbers.FB622.value}. Attribute `param_list` is empty. You should "
                 f"have initialized the model beforehand (try calling `set_init_params`)"
             )
+
+        local_params_set = set(local_params) if local_params else set()
+
+        unknown = local_params_set - set(self.param_list)
+        if unknown:
+            raise FedbiomedModelError(
```
Comment (sharkovsky) [inline fedbiomed/common/models/_sklearn.py:133]: it seems harsh to raise an exception here. 
Maybe simply log a warning, since it doesn't seem to me like the rest of the execution would fail if there are unknown parameters (but maybe I'm wrong, in which case it makes sense to raise an exception here)

File: fedbiomed/common/models/_torch.py [hunk 7]
```diff
@@ -66,25 +69,39 @@ def get_weights(
...
+
+        param_iterator = list(
             self.model.named_parameters()
             if exclude_buffers
             else self.model.state_dict().items()
         )
+
+        unknown = local_params_set - {name for name, _ in param_iterator}
+        if unknown:
+            raise FedbiomedModelError(
```
Comment (sharkovsky) [inline fedbiomed/common/models/_torch.py:88]: same comment as before for exception

File: fedbiomed/common/training_plans/_base_training_plan.py [hunk 8]
```diff
@@ -671,6 +698,117 @@ def evaluate(data, target):
...
+        Args:
+            name: The name of the parameter whose tags are being retrieved.
+
+        Returns:
+            The validated set of tags associated with the parameter.
+        """
+        tags = self.tag_parameters(name)
+        return self._validate_parameter_tags(name, tags)
+
+    def filter_model_params_by_tags(
```
Comment (sharkovsky) [inline fedbiomed/common/training_plans/_base_training_plan.py:773]: this should be a static method

**Comment by sharkovsky** [comment]
Great work @lchambon !

I've added some review comments, when fixed I think this is ready to be merged. 
I confirm the PR implements the intended behaviour.

I have a general comment about the implementation. Currently the interface for handling model parameters through the training plan (and through the underlying Model object) is becoming more and mode complex. 
It was already quite complex with all the `only_trainable` and `exclude_buffers` logic, but now with `filter_model_params_by_tags` it becomes necessary to think about whether we need to filter the model parameters every time we use them. 

The current implementation follows what was done before and fits within the rest of the state of affairs, however I think a better design and refactor is becoming overdue for this part. I will open a separate issue for that.

File: docs/user-guide/researcher/training-plan.md (hunk 4 above)
Comment (lchambon) [reply docs/user-guide/researcher/training-plan.md]: Thank you Francesco, I applied this change

File: docs/user-guide/researcher/training-plan.md (hunk 5 above)
Comment (lchambon) [reply docs/user-guide/researcher/training-plan.md:486]: Thank you Francesco, I added this paragraph:

```markdown
From the researcher's perspective, this has the following practical implications:

- `local` parameters are absent from aggregation results: they will not appear in `experiment.aggregated_params()`, nor in the raw per-node replies accessible via `experiment.training_replies()`.
- `local` parameters are still present in the researcher-side training plan: they remain accessible through `experiment.training_plan().get_model_params()` and `experiment.training_plan().model()`. However, these values originate from `init_model`: they reflect the initial weights as defined by the researcher, not the values that were actually trained and retained on each node. In particular, they do not correspond to the local parameter values that nodes will restore at the start of the next round (which, for parameters tagged as both `local` and `persistent`, are the node's own saved values from the previous round).
```

File: fedbiomed/common/models/_sklearn.py (hunk 6 above)
Comment (lchambon) [reply fedbiomed/common/models/_sklearn.py]: You are totally right. As suggested, I changed to log a warning instead:

```python
if unknown:
    logger.warning(
        "'BaseSkLearnModel.get_weights' received unknown local parameters "
        "that are not part of the model; unknown parameters: %s. ",
        unknown,
    )
```

File: fedbiomed/common/models/_torch.py (hunk 7 above)
Comment (lchambon) [reply fedbiomed/common/models/_torch.py]: Same for _sklearn, I changed to log a warning instead:

```python
if unknown:
    logger.warning(
        "'TorchModel.get_weights' received unknown local parameters "
        "that are not part of the model; unknown parameters: %s. ",
        unknown,
    )
```

File: fedbiomed/node/round.py (hunk 3 above)
Comment (lchambon) [reply fedbiomed/node/round.py]: Thank you Francesco, you are totally right ! I did not think about it. I double-checked and this works perfectly fine. 

I have replaced the function by:

```python
def _reconstruct_full_params(
    self,
    initial_model_weights: Dict[str, Any],
    weights_from_researcher: Dict[str, Any],
    persistent_model_weights: Optional[Dict[str, Any]],
) -> Dict[str, Any]:
    """Reconstructs the full set of model parameters before training.

    This method merges parameters received from the researcher (typically
    aggregated global parameters) with locally managed parameters based on
    their associated tags ("local", "persistent").

    The reconstruction follows these rules:
        - Parameters tagged as both "local" and "persistent" are restored
        from `persistent_model_weights` if available. If not (e.g., first
        round), they are initialized from `initial_model_weights`.
        - Parameters tagged as "local" but not "persistent" are always
        reinitialized from `initial_model_weights` at each round.
        - All other parameters are kept as provided in `weights_from_researcher`.

    Args:
        initial_model_weights: initial values of all model parameters (before any training)
        weights_from_researcher: parameters received from the researcher
        persistent_model_weights: locally stored parameters saved from previous rounds

    Returns:
        the complete set of parameters to be loaded into the model before training
    """
    # Deal with the case where persistent_model_weights is None
    persistent_model_weights = (
        persistent_model_weights if persistent_model_weights is not None else {}
        )

    return {**initial_model_weights, **persistent_model_weights, **weights_from_researcher}
```

File: fedbiomed/node/round.py (hunk 2 above)
Comment (lchambon) [reply fedbiomed/node/round.py]: It was to create a shallow copy of `weights_from_researcher`. But it is useless now, as I simplified a lot the function thanks to your other comment.

File: fedbiomed/node/round.py (hunk 1 above)
Comment (lchambon) [reply fedbiomed/node/round.py]: Thank you, you are right. I corrected as suggested.

File: fedbiomed/common/training_plans/_base_training_plan.py (hunk 8 above)
Comment (lchambon) [reply fedbiomed/common/training_plans/_base_training_plan.py:773]: I am a bit confused about this comment.

Do you mean doing something like this:

```python

@staticmethod
def filter_model_params_by_tags(
    params: Dict[str, Any],
    get_tags: Callable[[str], Set[str]],
    required_tags: Optional[Set[str]] = None,
    forbidden_tags: Optional[Set[str]] = None,
) -> Dict[str, Any]:
    """Filters a given parameter dict based on required or forbidden tags.

    Args:
        params: Dictionary of model parameters (e.g. output of get_model_params)
        get_tags: Function that returns tags for a given parameter name
        required_tags: Set of tags that parameters must include (AND logic)
        forbidden_tags: Set of tags that parameters must NOT include

    Returns:
        Filtered dictionary of parameters
    """
    filtered_params = {}

    required_tags = required_tags or set()
    forbidden_tags = forbidden_tags or set()

    if not isinstance(required_tags, set) or not isinstance(forbidden_tags, set):
        raise FedbiomedTrainingPlanError(
            f"{ErrorNumbers.FB605}: required_tags and forbidden_tags must be sets"
        )

    for name, value in params.items():
        tags = get_tags(name)

        if required_tags and not required_tags.issubset(tags):
            continue

        if forbidden_tags and not tags.isdisjoint(forbidden_tags):
            continue

        filtered_params[name] = value

    return filtered_params
````

And then using in `round.py` like this:

```python
BaseTrainingPlan.filter_model_params_by_tags(...)
```

in order to separate how to get tags from how to filter parameters ?

**Comment by lchambon** [comment]
Thank you @sharkovsky for your review. I applied your suggested changes, except for one comment I’d like to discuss with you, if you don’t mind.
I also added end2end tests if you want to review them.

File: fedbiomed/common/training_plans/_base_training_plan.py (hunk 8 above)
Comment (sharkovsky) [reply fedbiomed/common/training_plans/_base_training_plan.py:773]: Ah, forget what I said. I think I missed the fact that you were using `self.get_parameter_tags` inside the function. Forget what I said!
